lib/prism/parse_result.rb

```diff
@@ -108,16 +114,46 @@ class Location
     # The length of this location in bytes.
     attr_reader :length
 
-    # The list of comments attached to this location
-    attr_reader :comments
-
     # Create a new location object with the given source, start byte offset, and
     # byte length.
     def initialize(source, start_offset, length)
       @source = source
       @start_offset = start_offset
       @length = length
-      @comments = []
+
+      # These are used to store comments that are associated with this location.
+      # They are initialized to `nil` to save on memory when there are no
+      # comments to be attached and/or the comment-related APIs are not used.
+      @leading_comments = nil
+      @trailing_comments = nil
```

Comment:
I think in terms of memory usage, it's best to not assign these at all.
Assigning, even to `nil`, can cause extra memory to be used, because it's considered like any other value and e.g. the Shape will have `@leading_comments at OFFSET` and then in values[OFFSET] we'll have that `nil`.

Reply:
This is going to have a consistent shape, so the callsites to `leading_comments`, `trailing_comments`, and `comments` don't become polymorphic, depending on call order. The `nil` itself is just going to be a tagged pointer, so it'll be a minimal impact.

Reply:
Right, a monomorphic shape is also important.

For curiosity I measured it on `ruby 3.3.0 (2023-12-25 revision 5124f9ac75) [x86_64-linux]` and there is a difference:
```
with this change:
$ bundle exec ruby -Ilib -rprism -robjspace -e 'loc = Prism.parse("1").value.location; p loc; p ObjectSpace.memsize_of(loc)'
#<Prism::Location @start_offset=0 @length=1 start_line=1>
80

without these 2 lines:
$ bundle exec ruby -Ilib -rprism -robjspace -e 'loc = Prism.parse("1").value.location; p loc; p ObjectSpace.memsize_of(loc)'
#<Prism::Location @start_offset=0 @length=1 start_line=1>
40
```

So it's a trade-off between memory footprint and monomorphic shape.
Given Location objects are lazy now the footprint likely matters less.